src/engraving/libmscore/textedit.h

```diff
@@ -60,13 +60,75 @@ struct TextEditData : public ElementEditData {
 class TextEditUndoCommand : public UndoCommand
 {
 protected:
-    TextCursor c;
+    TextCursor _cursor;
 public:
     TextEditUndoCommand(const TextCursor& tc)
-        : c(tc) {}
+        : _cursor(tc) {}
     bool isFiltered(UndoCommand::Filter f, const Element* target) const override
     {
-        return f == UndoCommand::Filter::TextEdit && c.text() == target;
+        return f == UndoCommand::Filter::TextEdit && _cursor.text() == target;
+    }
+
+    TextCursor& cursor() { return _cursor; }
+};
+
+//---------------------------------------------------------
+//   ChangeText
+//---------------------------------------------------------
+
+class ChangeTextProperties : public TextEditUndoCommand
+{
+    QString xmlText;
+    Pid propertyId;
+    QVariant propertyVal;
+    FontStyle existingStyle;
+
+    void restoreSelection()
+    {
+        TextCursor& tc = cursor();
+        tc.text()->cursor()->setSelectLine(tc.selectLine());
+        tc.text()->cursor()->setSelectColumn(tc.selectColumn());
+        tc.text()->cursor()->setRow(tc.row());
+        tc.text()->cursor()->setColumn(tc.column());
+    }
+
+public:
+    ChangeTextProperties(const TextCursor* tc, Ms::Pid propId, const QVariant& propVal)
+        : TextEditUndoCommand(*tc)
+    {
+        propertyId = propId;
+        propertyVal = propVal;
+        if (propertyId == Pid::FONT_STYLE) {
+            existingStyle = static_cast<FontStyle>(cursor().text()->getProperty(propId).toInt());
+        }
+    }
+
+    virtual void undo(EditData*) override
+    {
+        cursor().text()->resetFormatting();
+        cursor().text()->setXmlText(xmlText);
+        restoreSelection();
+        cursor().text()->layout();
+    }
+
+    virtual void redo(EditData*) override
```

Comment:
1. better to move all implementation to cpp. We should leave only declarations in h, all implementation must be stored in cpp

2. you can remove virtual, because you already used override

Reply:
agree, just that everything else in here was done this way

Reply:
moved implementations to .cpp file